### PR TITLE
Add better version selection logic for `asoctl clean crds`

### LINKED ISSUE #2849: asoctl crd clean should have better version selection logic
See [PR](https://github.com/Azure/azure-service-operator/pull/2846#discussion_r1160303245) for more context.

The algorithm I think we actually need is:
1. Use storage version to list all CRs. Extract the OriginalGVK field
2.  Swap v1alpha1 -> v1beta1 (for alpha deprecation) and save that as versionToUse for that CR
3.  Issue GET + PUT with versionToUse

### PR DESCRIPTION
**Pull request by super-harsh**
Closes #2849 

**What this PR does / why we need it**:

In this PR, we have a better version selection logic. Where the cleaner lists storage resources using the latest storage versions and updates the original version from `v1alpha1api` to `v1beta`

**Special notes for your reviewer**:

Tested scenario:
alpha.6 -> apply rg -> apply storageAccount -> upgrade to beta.5 -> run `asoctl clean crds` -> upgrade to 2.0.0. 

In the above scenario, everything went smooth, except originalVersion for ResourceGroup was not found. However, the resources were present and successfully reconciled.

### CODE REVIEW

File: v2/tools/asoctl/internal/crd/cleaner.go [hunk 1]
```diff
@@ -236,9 +252,3 @@ func removeMatchingStoredVersions(oldVersions []string, versionRegexp *regexp.Re
 
 	return newStoredVersions, matchedStoredVersion
 }
-
-// getVersionFromStoredVersion returns the public (non-storage) API version for a given version
-func getVersionFromStoredVersion(version string) string {
```
Comment (matthchr) [inline v2/tools/asoctl/internal/crd/cleaner.go:241]: I think you actually do need to keep this for my comment on line 179.

File: v2/tools/asoctl/internal/crd/cleaner.go [hunk 2]
```diff
@@ -170,7 +162,31 @@ func (c *Cleaner) migrateObjects(ctx context.Context, objectsToMigrate *unstruct
 			continue
 		}
 
-		err := retry.OnError(c.migrationBackoff, isErrorFatal, func() error { return c.client.Update(ctx, &obj) })
+		originalVersionFieldPath := []string{"spec", "originalVersion"}
+
+		originalVersion, found, err := unstructured.NestedString(obj.Object, originalVersionFieldPath...)
+		if err != nil {
+			return errors.Wrap(err,
+				fmt.Sprintf("error while migrating %q of kind %s", obj.GetName(), obj.GroupVersionKind().Kind))
```
Comment (theunrepentantgeek) [inline v2/tools/asoctl/internal/crd/cleaner.go:170]: It's already an error, don't need to include that in the message.
```suggestion
				fmt.Sprintf("migrating %q of kind %s", obj.GetName(), obj.GroupVersionKind().Kind))
```

File: v2/tools/asoctl/internal/crd/cleaner.go [hunk 3]
```diff
@@ -170,7 +162,33 @@ func (c *Cleaner) migrateObjects(ctx context.Context, objectsToMigrate *unstruct
...
+			c.log.Info(
+				"originalVersion not found. Continuing with the latest.",
+				"name", obj.GetName(),
+				"kind", obj.GroupVersionKind().Kind)
+			// Here we set the originalVersion as active version since we can't find it
+			originalVersion = getVersionFromStoredVersion(activeVersion)
+		}
+
+		originalVersion = strings.Replace(originalVersion, "v1alpha1api", "v1beta", 1)
+		err = unstructured.SetNestedField(obj.Object, originalVersion, originalVersionFieldPath...)
```
Comment (matthchr) [inline v2/tools/asoctl/internal/crd/cleaner.go:185]: Should you only do this if `found`? Since if there wasnt' one to begin with you can't set it?

**Review by matthchr — APPROVED** [review]

File: v2/tools/asoctl/internal/crd/cleaner.go [hunk 4]
```diff
@@ -170,7 +162,31 @@ func (c *Cleaner) migrateObjects(ctx context.Context, objectsToMigrate *unstruct
...
+				fmt.Sprintf("error while migrating %q of kind %s", obj.GetName(), obj.GroupVersionKind().Kind))
+		}
+
+		// TODO: We continue here with the activeVersion
+		if !found {
+			c.log.Info("originalVersion not found. Continuing with the latest.",
+				"name", obj.GetName(),
+				"kind", obj.GroupVersionKind().Kind)
+			// Here we set the originalVersion as active version since we can't find it
+			originalVersion = activeVersion
```
[the remark this answers is not in the document]
Comment (super-harsh) [reply v2/tools/asoctl/internal/crd/cleaner.go]: Removed this bit as whole since we don't need to set version when OriginalVersion is not found.

**Pull request merged**